src/wasm/kotlin/java/arcs/wasm/WasmRuntimeClient.kt

```diff
@@ -52,30 +60,30 @@ actual object RuntimeClient {
     actual fun <T : Entity<T>> collectionStore(
         particle: Particle,
         collection: Collection<T>,
-        encoded: String
+        encoded: NullTermByteArray
     ): String? {
         val wasmId = collectionStore(
             particle.toAddress(),
             collection.toAddress(),
-            encoded.toWasmString()
+            encoded.bytes.toWasmAddress()
         )
         return wasmId.toNullableKString()?.let { _free(wasmId); it }
     }
 
-    actual fun log(msg: String) = arcs.wasm.log(msg);
+    actual fun log(msg: String) = arcs.wasm.log(msg)
 
-    actual fun onRenderOutput(particle: Particle, template: String?, model: String?) =
+    actual fun onRenderOutput(particle: Particle, template: String?, model: NullTermByteArray?) =
         onRenderOutput(
             particle.toAddress(),
             template.toWasmNullableString(),
-            model.toWasmNullableString()
+            model?.let { model.bytes.toWasmAddress() } ?: 0
```

Comment:
This is fine, because - given that the model variable is essentially a final value, it's is guaranteed to be non-null within the `let` block, but a better practice is to use `it` within the `let`. The `it` is guaranteed to be non-null even if the `model` pointer were to become null between the check and the body of the `let`.

```kotlin
model?.let { it.bytes.toWasmAddress() } ?: 0
```

But even better, you can do this:

```kotlin
model?.bytes?.toWasmAddress() ?: 0
```

Reply:
Changed to the latter, but I'm a bit puzzled as to why it works that way. `model` is nullable, so the first `?.` makes sense, but `NullTermByteArray.bytes` is *not* nullable, so given a valid model reference, why do we need to `?.` guard the bytes field?

Reply:
The return type of `model?.bytes` is nullable, because the `model` is nullable.

Reply:
Ok, so the first `?.` doesn't just shortcut the remaining expression? Given `a?.b.c.d` I would expect b, c, and d not to be evaluated when a is null.

Reply:
Breaking down `a?.b.c.d`:   

`a?.b` is fine, and returns a nullable value of b's type, but `a?.b.c` would be a null pointer exception since you're calling the `c` getter on a null object.

The compiler will stop you from doing `a?.b.c.d` and risking NPEs, but when evaluating - it *does* skip the remaining calls and jumps to the elvis (if there is one..)